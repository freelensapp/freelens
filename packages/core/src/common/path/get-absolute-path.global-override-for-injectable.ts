/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import path from "path";
import { getGlobalOverride } from "@freelensapp/test-utils";
import getAbsolutePathInjectable from "./get-absolute-path.injectable";

export default getGlobalOverride(getAbsolutePathInjectable, () => path.posix.resolve);
