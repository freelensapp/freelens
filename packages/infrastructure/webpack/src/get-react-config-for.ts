import MiniCssExtractPlugin from "mini-css-extract-plugin";
import type { Configuration } from "webpack";
import { type Paths, getNodeConfig } from "./get-node-config";

export const getReactConfigFor =
  ({ miniCssExtractPluginLoader = MiniCssExtractPlugin.loader } = {}) =>
  ({ entrypointFilePath, outputDirectory }: Paths): Configuration => {
    const nodeConfig = getNodeConfig({
      entrypointFilePath,
      outputDirectory,
    });

    return {
      ...nodeConfig,

      plugins: [
        ...nodeConfig.plugins!,

        new MiniCssExtractPlugin({
          filename: "[name].css",
        }),
      ],

      module: {
        ...nodeConfig.module,

        rules: [
          ...nodeConfig.module!.rules!,

          {
            test: /\.s?css$/,

            use: [
              miniCssExtractPluginLoader,

              {
                loader: "css-loader",

                options: {
                  sourceMap: false,

                  modules: {
                    auto: /\.module\./i, // https://github.com/webpack-contrib/css-loader#auto
                    mode: "local", // :local(.selector) by default
                    localIdentName: "[name]__[local]--[hash:base64:5]",
                  },
                },
              },

              {
                loader: "postcss-loader",
                options: {
                  sourceMap: false,
                  postcssOptions: {
                    plugins: ["tailwindcss"],
                  },
                },
              },

              {
                loader: "sass-loader",
                options: {
                  api: "modern",
                  sourceMap: false,
                },
              },
            ],
          },

          {
            test: /\.(ttf|eot|woff2?)$/,
            type: "asset/resource",
            generator: {
              filename: "fonts/[name][ext]",
            },
          },

          {
            test: /\.svg$/,
            type: "asset/source", // exports the source code of the asset, so we get XML
          },

          {
            test: /\.(jpg|png|ico)$/,
            type: "asset/resource",
            generator: {
              filename: "images/[name][ext]",
            },
          },
        ],
      },
    };
  };
