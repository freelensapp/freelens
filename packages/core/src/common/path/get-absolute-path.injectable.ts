/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import path from "path";
import { getInjectable } from "@ogre-tools/injectable";

export type GetAbsolutePath = (...args: string[]) => string;

const getAbsolutePathInjectable = getInjectable({
  id: "get-absolute-path",

  instantiate: (): GetAbsolutePath => path.resolve,

  // This causes side effect e.g. Windows creates different kinds of
  // absolute paths than linux
  causesSideEffects: true,
});

export default getAbsolutePathInjectable;
