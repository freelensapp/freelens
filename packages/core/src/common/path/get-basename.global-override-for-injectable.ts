/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import path from "path";
import { getGlobalOverride } from "@freelensapp/test-utils";
import getBasenameOfPathInjectable from "./get-basename.injectable";

export default getGlobalOverride(getBasenameOfPathInjectable, () => path.posix.basename);
