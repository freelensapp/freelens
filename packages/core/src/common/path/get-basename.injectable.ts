/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import path from "path";
import { getInjectable } from "@ogre-tools/injectable";

export type GetBasenameOfPath = (path: string) => string;

const getBasenameOfPathInjectable = getInjectable({
  id: "get-basename-of-path",
  instantiate: (): GetBasenameOfPath => path.basename,
  causesSideEffects: true,
});

export default getBasenameOfPathInjectable;
