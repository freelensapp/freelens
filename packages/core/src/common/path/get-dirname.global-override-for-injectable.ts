/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import path from "path";
import { getGlobalOverride } from "@freelensapp/test-utils";
import getDirnameOfPathInjectable from "./get-dirname.injectable";

export default getGlobalOverride(getDirnameOfPathInjectable, () => path.posix.dirname);
