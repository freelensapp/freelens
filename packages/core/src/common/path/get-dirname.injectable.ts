/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import path from "path";
import { getInjectable } from "@ogre-tools/injectable";

export type GetDirnameOfPath = (path: string) => string;

const getDirnameOfPathInjectable = getInjectable({
  id: "get-dirname-of-path",
  instantiate: (): GetDirnameOfPath => path.dirname,
  causesSideEffects: true,
});

export default getDirnameOfPathInjectable;
