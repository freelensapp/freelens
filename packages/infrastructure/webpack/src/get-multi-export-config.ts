import { filter, fromPairs, isEqual, keys, map, nth, reject, toPairs, values } from "lodash/fp";

import path from "path";
import { pipeline } from "@ogre-tools/fp";
import { getNodeConfig } from "./get-node-config";
import { getReactConfigFor } from "./get-react-config-for";

type Dependencies = {
  resolvePath: typeof path.resolve;
  workingDirectory: string;
  getReactConfig: ReturnType<typeof getReactConfigFor>;
};

export const getMultiExportConfig = (packageJson: any, _dependencies: Partial<Dependencies>) => {
  const dependencies: Dependencies = {
    resolvePath: path.resolve,
    workingDirectory: process.cwd(),
    getReactConfig: getReactConfigFor(),
    ..._dependencies,
  };

  if (!packageJson.lensMultiExportConfig) {
    throw new Error(`Tried to get multi export config for package "${packageJson.name}" but configuration is missing.`);
  }

  const validBuildTypes = ["node", "react"];

  const invalidBuildTypes = pipeline(
    packageJson.lensMultiExportConfig,
    values,
    map((config) => config.buildType),
    reject((buildType) => validBuildTypes.includes(buildType)),
  );

  if (invalidBuildTypes.length > 0) {
    throw new Error(
      `Tried to get multi export config for package "${packageJson.name}" but build types "${invalidBuildTypes.join(
        '", "',
      )}" were not any of "${validBuildTypes.join('", "')}".`,
    );
  }

  const exportsWithMissingEntrypoint = pipeline(
    packageJson.lensMultiExportConfig,
    toPairs,
    filter(([, config]) => !config.entrypoint),
    map(nth(0)),
  );

  if (exportsWithMissingEntrypoint.length > 0) {
    throw new Error(
      `Tried to get multi export config for package "${
        packageJson.name
      }" but entrypoint was missing for "${exportsWithMissingEntrypoint.join('", "')}".`,
    );
  }

  const expectedExports = pipeline(packageJson.lensMultiExportConfig, keys, map(toExpectedExport), fromPairs);

  if (!isEqual(expectedExports, packageJson.exports)) {
    throw new Error(
      `Tried to get multi export config but exports of package.json for "${
        packageJson.name
      }" did not match exactly:\n\n${JSON.stringify(expectedExports, null, 2)}`,
    );
  }

  const toExportSpecificWebpackConfig = toExportSpecificWebpackConfigFor(dependencies);

  return pipeline(packageJson.lensMultiExportConfig, toPairs, map(toExportSpecificWebpackConfig));
};

const toExpectedExport = (externalImportPath: string) => {
  const posixJoinForPackageJson = path.posix.join;

  const entrypointPath = `./${posixJoinForPackageJson("./dist", externalImportPath, "index.js")}`;

  return [
    externalImportPath,
    {
      types: `./${posixJoinForPackageJson("./dist", externalImportPath, "index.d.ts")}`,

      default: entrypointPath,
      import: entrypointPath,
      require: entrypointPath,
    },
  ];
};

const toExportSpecificWebpackConfigFor =
  (dependencies: Dependencies) =>
  ([externalImportPath, { buildType, entrypoint }]: [string, { buildType: string; entrypoint: string }]) => {
    const outputDirectory = dependencies.resolvePath!(dependencies.workingDirectory!, "dist", externalImportPath);

    return buildType === "node"
      ? getNodeConfig({
          entrypointFilePath: entrypoint,
          outputDirectory,
        })
      : dependencies.getReactConfig!({
          entrypointFilePath: entrypoint,
          outputDirectory,
        });
  };
