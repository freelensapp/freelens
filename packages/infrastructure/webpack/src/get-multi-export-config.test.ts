import path from "path";
import { inspect } from "util";
import ForkTsCheckerPlugin from "fork-ts-checker-webpack-plugin";
import type { Configuration } from "webpack";
import { getMultiExportConfig } from "./get-multi-export-config";
import { getReactConfigFor } from "./get-react-config-for";

const resolvePathFake = path.posix.resolve;

describe("get-multi-export-config", () => {
  let configs: Configuration[];
  let maximalPackageJson: any;

  beforeEach(() => {
    maximalPackageJson = {
      name: "some-name",

      lensMultiExportConfig: {
        ".": {
          buildType: "node",
          entrypoint: "./index.ts",
        },

        "./some-entrypoint": {
          buildType: "node",
          entrypoint: "./some-entrypoint/index.ts",
        },

        "./some-other-entrypoint": {
          buildType: "react",
          entrypoint: "./some-other-entrypoint/index.ts",
        },
      },

      exports: {
        ".": {
          types: "./dist/index.d.ts",
          require: "./dist/index.js",
          import: "./dist/index.js",
          default: "./dist/index.js",
        },
        "./some-entrypoint": {
          types: "./dist/some-entrypoint/index.d.ts",
          require: "./dist/some-entrypoint/index.js",
          import: "./dist/some-entrypoint/index.js",
          default: "./dist/some-entrypoint/index.js",
        },
        "./some-other-entrypoint": {
          types: "./dist/some-other-entrypoint/index.d.ts",
          require: "./dist/some-other-entrypoint/index.js",
          import: "./dist/some-other-entrypoint/index.js",
          default: "./dist/some-other-entrypoint/index.js",
        },
      },
    };
  });

  describe("given maximal package.json, when creating configuration", () => {
    beforeEach(() => {
      configs = getMultiExportConfig(maximalPackageJson, {
        resolvePath: resolvePathFake,
        workingDirectory: "/some-working-directory",

        getReactConfig: getReactConfigFor({
          miniCssExtractPluginLoader: "miniCssExtractPluginLoader",
        }),
      });
    });

    it("works", () => {
      expect(
        inspect(configs, {
          colors: false,
          depth: Infinity,
          maxArrayLength: Infinity,
          maxStringLength: Infinity,
        }),
      ).toMatchSnapshot();
    });

    [
      {
        name: "config for node export in default entrypoint",
        entrypoint: "./index.ts",
        outputDirectory: "/some-working-directory/dist",
      },
      {
        name: "config for node export in a non-default entrypoint",
        entrypoint: "./some-entrypoint/index.ts",
        outputDirectory: "/some-working-directory/dist/some-entrypoint",
      },
      {
        name: "config for react export in a non-default entrypoint",
        entrypoint: "./some-other-entrypoint/index.ts",
        outputDirectory: "/some-working-directory/dist/some-other-entrypoint",
      },
    ].forEach((scenario) => {
      describe(scenario.name, () => {
        let config: Configuration;

        beforeEach(() => {
          config = configs.find(({ name }) => name === scenario.entrypoint)!;
        });

        it("has correct entrypoint", () => {
          expect(config).toHaveProperty("entry.index", scenario.entrypoint);
        });

        it("has correct output directory", () => {
          expect(config).toHaveProperty("output.path", scenario.outputDirectory);
        });

        it("has correct declaration directory", () => {
          expect(config.plugins!.find((fn: any) => fn?.constructor === ForkTsCheckerPlugin)).toHaveProperty(
            "options.typescript.configOverwrite.compilerOptions.declarationDir",
            scenario.outputDirectory,
          );
        });
      });
    });
  });

  it("given maximal package.json but path for entrypoint in exports do not match output, when creating configuration, throws", () => {
    maximalPackageJson.exports["./some-entrypoint"].default = "wrong-path";

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        getReactConfig: () => reactConfigStub,
      });
    }).toThrow('Tried to get multi export config but exports of package.json for "some-name" did not match exactly:');
  });

  it("given maximal package.json but exports do not match lens multi export config, when creating configuration, throws", () => {
    maximalPackageJson.exports = {};

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        getReactConfig: () => reactConfigStub,
      });
    }).toThrow('Tried to get multi export config but exports of package.json for "some-name" did not match exactly:');
  });

  it("given maximal package.json but exports are missing, when creating configuration, throws", () => {
    delete maximalPackageJson.exports;

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        getReactConfig: () => reactConfigStub,
      });
    }).toThrow('Tried to get multi export config but exports of package.json for "some-name" did not match exactly:');
  });

  it("given maximal package.json but lens multi export config is missing, when creating configuration, throws", () => {
    delete maximalPackageJson.lensMultiExportConfig;

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        getReactConfig: () => reactConfigStub,
      });
    }).toThrow('Tried to get multi export config for package "some-name" but configuration is missing.');
  });

  it("given maximal package.json but a build type is incorrect, when creating configuration, throws", () => {
    maximalPackageJson.lensMultiExportConfig["."].buildType = "some-invalid";

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        getReactConfig: () => reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config for package "some-name" but build types "some-invalid" were not any of "node", "react".',
    );
  });

  it("given maximal package.json but entrypoint is missing, when creating configuration, throws", () => {
    delete maximalPackageJson.lensMultiExportConfig["./some-entrypoint"].entrypoint;

    expect(() => {
      getMultiExportConfig(maximalPackageJson, {
        getReactConfig: () => reactConfigStub,
      });
    }).toThrow(
      'Tried to get multi export config for package "some-name" but entrypoint was missing for "./some-entrypoint".',
    );
  });
});

const reactConfigStub = {
  stub: "react",

  output: {
    some: "other-value",
    path: "/some-build-directory",
  },
} as unknown as Configuration;
