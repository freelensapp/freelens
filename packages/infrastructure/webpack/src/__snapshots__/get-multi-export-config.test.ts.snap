// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-multi-export-config given maximal package.json, when creating configuration works 1`] = `
"[
  {
    name: './index.ts',
    entry: { index: './index.ts' },
    target: 'node',
    mode: 'production',
    performance: { maxEntrypointSize: 500000, hints: 'error' },
    resolve: { extensions: [ '.ts', '.tsx', '.js' ] },
    plugins: [
      MakePeerDependenciesExternalPlugin {},
      ProtectFromImportingNonDependencies {},
      ForkTsCheckerWebpackPlugin {
        options: {
          typescript: {
            mode: 'write-dts',
            configOverwrite: {
              include: [ './index.ts' ],
              compilerOptions: {
                declaration: true,
                declarationDir: '/some-working-directory/dist'
              }
            }
          }
        }
      }
    ],
    output: {
      path: '/some-working-directory/dist',
      filename: [Function: filename],
      library: { type: 'commonjs2' }
    },
    optimization: { minimize: false },
    externalsPresets: { node: true },
    externals: [ [Function (anonymous)] ],
    node: { __dirname: true, __filename: true },
    module: { rules: [ { test: /\\.ts(x)?$/, loader: 'ts-loader' } ] }
  },
  {
    name: './some-entrypoint/index.ts',
    entry: { index: './some-entrypoint/index.ts' },
    target: 'node',
    mode: 'production',
    performance: { maxEntrypointSize: 500000, hints: 'error' },
    resolve: { extensions: [ '.ts', '.tsx', '.js' ] },
    plugins: [
      MakePeerDependenciesExternalPlugin {},
      ProtectFromImportingNonDependencies {},
      ForkTsCheckerWebpackPlugin {
        options: {
          typescript: {
            mode: 'write-dts',
            configOverwrite: {
              include: [ './some-entrypoint/index.ts' ],
              compilerOptions: {
                declaration: true,
                declarationDir: '/some-working-directory/dist/some-entrypoint'
              }
            }
          }
        }
      }
    ],
    output: {
      path: '/some-working-directory/dist/some-entrypoint',
      filename: [Function: filename],
      library: { type: 'commonjs2' }
    },
    optimization: { minimize: false },
    externalsPresets: { node: true },
    externals: [ [Function (anonymous)] ],
    node: { __dirname: true, __filename: true },
    module: { rules: [ { test: /\\.ts(x)?$/, loader: 'ts-loader' } ] }
  },
  {
    name: './some-other-entrypoint/index.ts',
    entry: { index: './some-other-entrypoint/index.ts' },
    target: 'node',
    mode: 'production',
    performance: { maxEntrypointSize: 500000, hints: 'error' },
    resolve: { extensions: [ '.ts', '.tsx', '.js' ] },
    plugins: [
      MakePeerDependenciesExternalPlugin {},
      ProtectFromImportingNonDependencies {},
      ForkTsCheckerWebpackPlugin {
        options: {
          typescript: {
            mode: 'write-dts',
            configOverwrite: {
              include: [ './some-other-entrypoint/index.ts' ],
              compilerOptions: {
                declaration: true,
                declarationDir: '/some-working-directory/dist/some-other-entrypoint'
              }
            }
          }
        }
      },
      MiniCssExtractPlugin {
        _sortedModulesCache: WeakMap { <items unknown> },
        options: {
          filename: '[name].css',
          ignoreOrder: false,
          experimentalUseImportModule: undefined,
          runtime: true,
          chunkFilename: '[name].css'
        },
        runtimeOptions: {
          insert: undefined,
          linkType: 'text/css',
          attributes: undefined
        }
      }
    ],
    output: {
      path: '/some-working-directory/dist/some-other-entrypoint',
      filename: [Function: filename],
      library: { type: 'commonjs2' }
    },
    optimization: { minimize: false },
    externalsPresets: { node: true },
    externals: [ [Function (anonymous)] ],
    node: { __dirname: true, __filename: true },
    module: {
      rules: [
        { test: /\\.ts(x)?$/, loader: 'ts-loader' },
        {
          test: /\\.s?css$/,
          use: [
            'miniCssExtractPluginLoader',
            {
              loader: 'css-loader',
              options: {
                sourceMap: false,
                modules: {
                  auto: /\\.module\\./i,
                  mode: 'local',
                  localIdentName: '[name]__[local]--[hash:base64:5]'
                }
              }
            },
            {
              loader: 'postcss-loader',
              options: {
                sourceMap: false,
                postcssOptions: { plugins: [ 'tailwindcss' ] }
              }
            },
            { loader: 'sass-loader', options: { sourceMap: false } }
          ]
        },
        {
          test: /\\.(ttf|eot|woff2?)$/,
          type: 'asset/resource',
          generator: { filename: 'fonts/[name][ext]' }
        },
        { test: /\\.svg$/, type: 'asset/source' },
        {
          test: /\\.(jpg|png|ico)$/,
          type: 'asset/resource',
          generator: { filename: 'images/[name][ext]' }
        }
      ]
    }
  }
]"
`;
